### PR TITLE
Drop Python 3.9

### LINKED ISSUE #680: Drop support for Python 3.9
### Description

NA

### PR DESCRIPTION
**Pull request by Armavica**
## Related Issue

- [x] Closes #680 
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [x] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Review by maresb — APPROVED** [review]
Looks fine, but I'm wondering why? Is 3.9 that old??? :older_man:

**Comment by Armavica** [comment]
> Looks fine, but I'm wondering why? Is 3.9 that old??? 👴

This might be a reason: https://scientific-python.org/specs/spec-0000/

**Comment by ricardoV94** [comment]
> > Looks fine, but I'm wondering why? Is 3.9 that old??? 👴
> 
> This might be a reason: https://scientific-python.org/specs/spec-0000/

Yes that's the reason

**Comment by maresb** [comment]
Makes sense, though sometimes it feels like the world moves too quickly.

File: pytensor/tensor/basic.py [hunk 1]
```diff
@@ -1671,22 +1667,14 @@ def do_constant_folding(self, fgraph, node):
                 idx == 0
                 and isinstance(
                     client.op,
-                    (
-                        # Ops that will work inplace on the Alloc. So if they
```
Comment (twiecki) [inline pytensor/tensor/basic.py:1675]: I don't think it should remove thsi comment.

Comment (Armavica) [reply pytensor/tensor/basic.py:1675]: Indeed, incredible catch! I saw a few more like that, I will take a closer look.

Comment (Armavica) [reply pytensor/tensor/basic.py:1675]: Ok I got all the comments back, checked with `git diff main | rg '^[+-].*#'`

**Comment by ricardoV94** [comment]
I don't care, but the whole point of that spec is exactly that no scientific package has to support more than 3 major python versions

**Comment by maresb** [comment]
Sorry, I butchered the versions in my above comment (or I'm not thinking straight). If we merge this right now, we'll only be supporting 2 major Python versions.

With my suggestion, we'll have one version that supports 4 and then drop down to a nominal 3 as you state.

**Comment by ricardoV94** [comment]
This will be incorporated as a "major release", so PyMC will only be able to depend on it when we manually bump ot over there, so it can't have any effect?

**Comment by maresb** [comment]
Ah... ok, I think that makes sense.

You're saying that we merge this and release it as PyTensor 3.0. Then over in PyMC, completely independently we increase the range to include PyTensor 2.19 and support Python 3.9-3.12. Then, again in PyMC completely independently when we pin to PyTensor 3, we support Python 3.10-3.12.

**Review by maresb — APPROVED** [review]

**Comment by ricardoV94** [comment]
Yes except we are not using semver so major changes happen on intermediate numbers.

**Comment by ricardoV94** [comment]
PyPI / Build wheels on ubuntu-20.04 failed, I restarted it.

**Comment by ricardoV94** [comment]
Thanks @Armavica!

**Pull request merged**